zerostash/src/commands/mount.rs

```diff
@@ -0,0 +1,381 @@
+        let mut stack = self.stack.lock().unwrap();
+        stack.insert(real_path.to_path_buf(), ChunkDataStack::default());
+        Ok((0, 0))
+    }
+    fn read(
+        &self,
+        _req: RequestInfo,
+        path: &Path,
+        _fh: u64,
+        offset: u64,
+        size: u32,
+        callback: impl FnOnce(ResultSlice<'_>) -> CallbackResult,
+    ) -> CallbackResult {
+        debug!("read: {:?} {:#x} @ {:#x}", path, size, offset);
+
+        let real_path = strip_path(path);
+        let path_string = real_path.to_str().unwrap();
+        let metadata = self.stash.index().files.get(path_string).unwrap();
+        let file_size = metadata.size as usize;
+        let offset = offset as usize;
+
+        if offset > file_size {
+            return callback(Err(libc::EINVAL));
+        }
+
+        let size = size as usize;
+        let mut chunks = metadata.chunks.clone();
+        chunks.sort_by(|(a, _), (b, _)| a.cmp(b));
+        let mut objectreader = self.stash.storage_reader().unwrap();
+
+        {
+            let mut stack = self.stack.lock().unwrap();
+            let stack = stack
+                .entry(real_path.to_path_buf())
+                .or_insert_with(ChunkDataStack::default);
+
+            if stack.last_read_offset == offset {
+                let end = size.min(file_size - offset);
+                if stack.stack.len() < end {
+                    loop {
+                        if stack
+                            .add_chunks(&chunks, file_size, &mut objectreader)
+                            .is_err()
+                        {
+                            return callback(Err(libc::EINVAL));
+                        }
+                        if stack.stack.len() >= size.min(file_size - offset) {
+                            break;
+                        }
+                    }
+                }
+
+                let ret_buf = stack.split_buf(end);
+                stack.update_current_read(offset + end);
+                return callback(Ok(&ret_buf));
+            }
+        }
+
+        let mut buf = vec![];
+        let mut from = None;
+        for (i, (c_offset, pointer)) in chunks.iter().enumerate() {
+            let c_offset = *c_offset as usize;
+
+            let next_c_offset = get_next_chunk_offset(file_size, &chunks, i);
+
+            if !buf.is_empty() || offset < next_c_offset {
+                if from.is_none() {
+                    from = Some(offset - c_offset);
+                }
+                let mut temp_buf = vec![0; next_c_offset - c_offset];
+                objectreader.read_chunk(pointer, &mut temp_buf).unwrap();
+                buf.append(&mut temp_buf);
+            }
+
+            if let Some(from) = from {
+                if buf[from..].len() >= size.min(file_size - offset) {
+                    let to = buf.len().min(from + size);
+                    return callback(Ok(&buf[from..to]));
+                }
+            }
+        }
```

Comment:
Why does this avoid using the chunkstack cache? It would be helpful to see a comment about when we arrive at this code.